src/bosh-director/lib/bosh/director/agent_broadcaster.rb

```diff
@@ -102,9 +104,10 @@ def sync_dns(instances, blobstore_id, sha1, version)
     end
 
     def filter_instances(vm_cid_to_exclude)
-      Models::Instance
+      Config.db[:instances]
         .inner_join(:vms, Sequel.qualify('vms', 'instance_id') => Sequel.qualify('instances','id'))
         .select_append(Sequel.expr(Sequel.qualify('instances','id')).as(:id))
+        .select_append(Sequel.expr(Sequel.qualify('vms','agent_id')).as(:fetched_agent_id))
         .where { Sequel.expr(Sequel.qualify('vms','active') => true) }
```

Comment:
This line is what I'm trying to do: specifically add the `agent_id` to an `instance`, without having to load the `vm` for each `instance` with a separate `SELECT * FROM "vms" WHERE (("instance_id" = 3683) AND ("active" IS TRUE)) LIMIT 1`

I'm not sure why the query in place is already using a lot of sequel internals like specific `inner_join`, `qualify`, and `select_append`, it seems pretty much all of it could be done with a single `eager_graph` like `Artist.eager_graph(albums: proc{|ds| ds.where{year > 1990}}).all`

I'll give that a try and see what happens, thanks!

Reply:
@charleshansen ok, I give up. I tried wrestling sequel into what's necessary, but couldn't do it. Here is what I'm unsure how this would work in sequel
* `instance` has a `one_to_many` with `vms`
* we want to achieve that for an `instance` `i.agent_id` *does not lazily do `SELECT * FROM "vms" WHERE (("instance_id" = 3683) AND ("active" IS TRUE)) LIMIT 1`
* At the same time, the above method needs to filter out instances based on fields in the associated `vm`. This is what I was unable to figure out in sequel. Any claused passed to `eager_graph` only filters on what to eagerly load, not what to keep from the list of instances.
  * examples are: a `vm.cid` which gets passed into the method and VMs with `vm.active != true` 

Not sure how/if that is possible in sequel.

Reply:
I looked into the query with @xtreme-behrouz-soroushian 

The existing query is more complicated than it needs to be, doing sql calculations across multiple tables. It should actually be both faster and simpler to do the filtering in memory instead of using sql queries. Here is our crack at it (all the unit tests pass)
```
    def filter_instances(vm_cid_to_exclude)
      instances_from_db = Models::Instance
        .where(compilation: false)
        .eager(:vms)
        .all

      instances_from_db.select do |instance|
        instance.vms.any? do |vm|
          vm.active && vm.cid != vm_cid_to_exclude
        end
      end
    end 
```

One complicating factor: `instance.agent_id` goes through [instance.active_vm](https://github.com/cloudfoundry/bosh/blob/e163565666c03e81e72481c630ecf1491b827039/src/bosh-director/lib/bosh/director/models/instance.rb#L146-L148) which appears to go around anything we could do with eager loading. If `active_vm` used (self.)`vms.find` instead of `Vm.first`, it would work, but it has slightly different caching behavior and I'm not sure if this is a safe change to make.

Reply:
We looked at it a little more, and tried to fix the `active_vm` reloading behavior (https://github.com/cloudfoundry/bosh/commit/e53000862b4df58c7e7442125be202a967780f83), but this breaks a number of integration tests.

My recommendation at this point is to directly use `instance.vms.find(&:active).agent_id` instead of `instance.agent_id` if you need caching behavior. It might be worth it to add an instance method to Models::Instance, something like `cached_agent_id` or `cached_active_vm`. If this doesn't work, let me know.

Reply:
> It should actually be both faster and simpler to do the filtering in memory instead of using sql queries.

This might sound stupid, but how did you validate this?

Reply:
On master, we changed `filter_instances` to the function we wrote above and the unit tests all passed. We got the query slightly wrong originally and the tests caught it, so I'm pretty sure this is ultimately the same function. I personally find it much easier to understand without all the extra sql logic.

As far as validating "faster", we didn't run any performance tests or anything, it was just more of a gut feeling. I've found in optimizing other queries that Ruby in memory is normally just as fast as sql unless you are using really large tables (~million rows) or if you are searching for something and utilizing database indexing. We aren't indexing on the "active" field in the vms, and I've never heard of anything approaching a million VMS in a bosh deployment, so it seems safe to me. 

That said, If you have a performance environment set up, I'd like to know how it fares. You would still need to stop using `instance.agent_id` so that we don't hit the database for every instance. I didn't try running it, so I'm not 100% sure this will fix it.